### PR TITLE
:bug: Allow external redis/rabbitmq secret creation even if the broke…

### PR DESCRIPTION
**Pull request by afreyermuth98**
# What this PR does
This PR allows the chart to create the secret of your redis / rabbitmq even if it's not the broker.
Actually, this is blocking if we want to have a redis as cache and a rabbitmq as broker for example

## Which issue(s) this PR fixes
Closes https://github.com/grafana/oncall/issues/2979

## Checklist

- [x] Unit, integration, and e2e (if applicable) tests updated
- [x] Documentation added (or `pr:no public docs` PR label added if not required)
- [x] `CHANGELOG.md` updated (or `pr:no changelog` PR label added if not required)

### CODE REVIEW

**Review by joeyorlando — COMMENTED** [review]
hi there @afreyermuth98 👋 

Thanks for your contribution! Do you mind adding two things:
- a few helm unit tests which assert the scenario you are trying to fix
- add an entry to the root `CHANGELOG.md` under (Unreleased > Fixed)

Thank you! 👊

**Comment by afreyermuth98** [comment]
Hey ! 
It's done :D
Btw I've modified a bit my code as we only use rabbitmq as a broker I guess :)

**Comment by joeyorlando** [comment]
@afreyermuth98 lastly you just need to sign the CLA 🙂 

https://github.com/grafana/oncall/pull/3903#issuecomment-1945687387

**Comment by afreyermuth98** [comment]
@joeyorlando Hmm weird i've signed it just after my commit 🤔

**Comment by joeyorlando** [comment]
> @joeyorlando Hmm weird i've signed it just after my commit 🤔 
> 
> ```
> 
> You have agreed to the CLA for grafana/oncall
> 
> ```

looks to be fine now, thank you!

File: CHANGELOG.md [hunk 1]
```diff
@@ -19,6 +19,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Unblock slack install by skipping check chatops gateway link in OSS deployment @mderynck ([#3893](https://github.com/grafana/oncall/pull/3893))
 - Fix multiple issues of alert groups table ([#3894](https://github.com/grafana/oncall/issues/3894))
 - Improvements for dragging the add rotation form in Schedules ([#3904](https://github.com/grafana/oncall/pull/3904))
+- Allow external-redis secret creation when the broker is rabbitmq ([#3903]https://github.com/grafana/oncall/pull/3903)
```
Comment (joeyorlando) [inline CHANGELOG.md:22]:
```suggestion
- Allow external-redis secret creation when the broker is rabbitmq ([#3903](https://github.com/grafana/oncall/pull/3903))
```

**Comment by joeyorlando** [comment]
@afreyermuth98 one last thing. One of the new helm unit tests is failing:

https://github.com/grafana/oncall/actions/runs/7971753982/job/21762105142#step:5:1

**Comment by afreyermuth98** [comment]
@joeyorlando Yeah I don't really understand why. The test fails because the password is not set but it's the normal functioning isn't it ? If I'm wrong I would be glad to be highlighted :)

File: helm/oncall/tests/broker_secret_test.yaml
Comment (joeyorlando) [inline helm/oncall/tests/broker_secret_test.yaml:1]: what about the case where externalRedis and externalRabbitmq are set?

**Comment by afreyermuth98** [comment]
Hey @joeyorlando I've modified the tests. It passed but it's not perfect and I would be glad to be highlighted.
Firstly, I had to comment my last lines of the `broker_secret_test.yaml` and I can't understand why it doesn't pass with.
Also, on `secrets.yaml - L64` , I've put `(.Values.externalRedis.host)` as a condition to pass (as you put the host for sure if you set an externalRedis) but I would have prefered to have `(.Values.externalRedis.password)` as before but I've investigated for hours and I can't understand why the test doesn't pass, everything is ok for me. 
Thanks by advance !

**Comment by joeyorlando** [comment]
> Hey @joeyorlando I've modified the tests. It passed but it's not perfect and I would be glad to be highlighted. Firstly, I had to comment my last lines of the `broker_secret_test.yaml` and I can't understand why it doesn't pass with. Also, on `secrets.yaml - L64` , I've put `(.Values.externalRedis.host)` as a condition to pass (as you put the host for sure if you set an externalRedis) but I would have prefered to have `(.Values.externalRedis.password)` as before but I've investigated for hours and I can't understand why the test doesn't pass, everything is ok for me. Thanks by advance !

One trick I just figured out is if you add [a `matchSnapshot` assertion](https://github.com/helm-unittest/helm-unittest?tab=readme-ov-file#snapshot-testing) it will spit out the rendered `.yaml` as such:
```yaml
# test file
- it: externalRedis.password and externalRabbitmq.password -> should create secret -redis-external and -rabbitmq-external
    templates:
      - engine/deployment.yaml
      - celery/deployment.yaml
    set:
      rabbitmq.enabled: false
      redis.enabled: false
      broker.type: rabbitmq
      externalRedis:
        host: redis.example.com
        username: user123
        password: abcd123
      externalRabbitmq:
        host: custom-host
        user: custom-user
        password: custom-password
    asserts:
      ... other assertions
      - matchSnapshot: {}
         template: secrets.yaml
```

then in `./helm/oncall/tests/__snapshot__` you will have a `broker_secret_test.yaml.snap` as such:
```yaml
externalRedis.password and externalRabbitmq.password -> should create secret -redis-external and -rabbitmq-external:
  1: |
    apiVersion: v1
    data:
      MIRAGE_SECRET_KEY: aEV4RmNqYm5sdllPVUtqVmZ2ZG1oc3N4NXRQNGpuc0E2dzV5eXJrOA==
      SECRET_KEY: WjNGRVUzak9IMkM3TTBiQ0dyZUpDTmZ5dnQ0aEticGFFaEdXUk5IcQ==
    kind: Secret
    metadata:
      labels:
        app.kubernetes.io/instance: oncall
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/name: oncall
        app.kubernetes.io/version: v1.3.99
        helm.sh/chart: oncall-1.3.99
      name: oncall
    type: Opaque
  2: |
    apiVersion: v1
    data:
      rabbitmq-password: Y3VzdG9tLXBhc3N3b3Jk
    kind: Secret
    metadata:
      name: oncall-rabbitmq-external
    type: Opaque
  3: |
    apiVersion: v1
    data:
      redis-password: YWJjZDEyMw==
    kind: Secret
    metadata:
      name: oncall-redis-external
    type: Opaque
```

turns out it was simply the `documentIndex` on the test you had commented out. It should be "2" instead of "1" 😄

**Comment by afreyermuth98** [comment]
Ooook I was wondering about this documentIndex but didn't find a conclusion about it ahah. Thanks a lot ! :D

**Pull request merged**